diff --git a/configure.py b/configure.py
index 03b5ce4..95f27aa 100755
--- a/configure.py
+++ b/configure.py
@@ -65,12 +65,14 @@ class Platform(object):
             self._platform = 'os400'
         elif self._platform.startswith('dragonfly'):
             self._platform = 'dragonfly'
+        elif self._platform.startswith('zos'):
+            self._platform = 'zos'
 
     @staticmethod
     def known_platforms() -> List[str]:
       return ['linux', 'darwin', 'freebsd', 'openbsd', 'solaris', 'sunos5',
               'mingw', 'msvc', 'gnukfreebsd', 'bitrig', 'netbsd', 'aix',
-              'dragonfly']
+              'dragonfly', 'zos']
 
     def platform(self) -> str:
         return self._platform  # type: ignore # Incompatible return value type
@@ -103,6 +105,9 @@ class Platform(object):
     def is_os400_pase(self) -> bool:
         return self._platform == 'os400' or os.uname().sysname.startswith('OS400')  # type: ignore # Module has no attribute "uname"
 
+    def is_zos(self) -> bool:
+        return self._platform == 'zos'
+
     def uses_usr_local(self) -> bool:
         return self._platform in ('freebsd', 'openbsd', 'bitrig', 'dragonfly', 'netbsd')
 
@@ -281,7 +286,11 @@ if configure_env:
     n.variable('configure_env', config_str + '$ ')
 n.newline()
 
-CXX = configure_env.get('CXX', 'c++')
+if platform.is_zos():
+    CXX = configure_env.get('CXX', 'clang++')
+else:
+    CXX = configure_env.get('CXX', 'c++')
+
 objext = '.o'
 if platform.is_msvc():
     CXX = 'cl'
@@ -359,6 +368,20 @@ if platform.is_msvc():
     if not options.debug:
         cflags += ['/Ox', '/DNDEBUG', '/GL']
         ldflags += ['/LTCG', '/OPT:REF', '/OPT:ICF']
+elif platform.is_zos():
+    cflags = ['-Wall', '-Wextra',
+       '-Wno-deprecated',
+       '-Wno-missing-field-initializers',
+       '-Wno-unused-parameter',
+       '-D_XPLATFORM_SOURCE',
+       '-D_POSIX_SOURCE',
+       '-D_OPEN_SYS_FILE_EXT',
+       '-DNINJA_PYTHON="%s"' % options.with_python]
+    if options.debug:
+       cflags += ['-g', '-D_GLIBCXX_DEBUG', '-D_GLIBCXX_DEBUG_PEDANTIC']
+    else:
+       cflags += ['-O2', '-DNDEBUG']
+    ldflags = ['-L$builddir']
 else:
     cflags = ['-g', '-Wall', '-Wextra',
               '-Wno-deprecated',
@@ -408,6 +431,8 @@ elif platform.is_aix():
     cflags.remove('-fvisibility=hidden')
 elif platform.is_msvc():
     pass
+elif platform.is_zos():
+    pass
 else:
     if options.profile == 'gmon':
         cflags.append('-pg')
@@ -504,7 +529,10 @@ if platform.supports_ninja_browse():
 n.comment('the depfile parser and ninja lexers are generated using re2c.')
 def has_re2c() -> bool:
     try:
-        proc = subprocess.Popen(['re2c', '-V'], stdout=subprocess.PIPE)
+        if platform.is_zos():
+            proc = subprocess.Popen(['re2c', '-V'], stdout=subprocess.DEVNULL)
+        else:
+            proc = subprocess.Popen(['re2c', '-V'], stdout=subprocess.PIPE)
         return int(proc.communicate()[0], 10) >= 1503
     except OSError:
         return False
@@ -580,6 +608,8 @@ else:
         objs += cxx(name, variables=cxxvariables)
 if platform.is_aix():
     objs += cc('getopt')
+elif platform.is_zos():
+    objs += cc('getopt')
 if platform.is_msvc():
     ninja_lib = n.build(built('ninja.lib'), 'ar', objs)
 else:
